[GTK][WPE] Optimize the case of adding Damage to an empty Damage

carlosgcampos opened the pull request:
#### 9993e5909f5124e3c2e5873d12d3479fe38413de
<pre>
[GTK][WPE] Optimize the case of adding Damage to an empty Damage
<a href="https://bugs.webkit.org/show_bug.cgi?id=290613">https://bugs.webkit.org/show_bug.cgi?id=290613</a>

Reviewed by Alejandro G. Castro.

In that case we can directly unite the given damage rects. We can also
optimize the case of unite() when there&apos;s only one rect, since we
already have the unite in m_minimumBoundingRectangle we can just copy it
to the first vector position.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::add):
(WebCore::Damage::cellIndexForRect const):
(WebCore::Damage::unite):
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, AddRects)):

Canonical link: <a href="https://commits.webkit.org/293001@main">https://commits.webkit.org/293001@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a43df602bccd193b357e92a00a815758f2b7fd4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16988 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7204 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102450 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47891 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99408 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17281 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25477 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74171 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31356 "Found 1 new test failure: workers/worker-to-worker.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100366 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13069 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88045 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54514 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12832 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5930 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47334 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82864 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6011 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104470 "Built successfully") | 
| | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24442 "Hash 2a43df60 for PR 43182 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17847 "Found 2 new test failures: imported/w3c/web-platform-tests/workers/data-url-shared.html ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83215 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24813 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84174 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82636 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27181 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4868 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17986 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15787 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24404 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29572 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24226 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27540 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25800 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -206,17 +206,52 @@ class Damage {
+    {
+        if (other.isEmpty() || !shouldAdd())
+            return false;
+
```
A note for the future.

Here, if `*this` and `other` both have the same `m_rect` and both `m_shouldUnite==true`, we can add another fastpath that would just iterate the `i` from `0` to `m_rects.size()` and just `m_rects[i].unite(other.m_rects[i])` as in that case both damage objects have the same layout of grid-powered rectangles.

The above fastpath would be useful in case of `AcceleratedSurfaceDMABuf::RenderTarget::addDamage` where usually all the conditions are met

carlosgcampos replied:
Let's do it in this patch then. Thanks!

Scony reviewed (commented):
LGTM

alexgcastro reviewed (approved):
LGTM!

carlosgcampos committed:
[GTK][WPE] Optimize the case of adding Damage to an empty Damage
https://bugs.webkit.org/show_bug.cgi?id=290613

Reviewed by Alejandro G. Castro.

In that case we can directly unite the given damage rects. We can also
optimize the case of unite() when there's only one rect, since we
already have the unite in m_minimumBoundingRectangle we can just copy it
to the first vector position.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::add):
(WebCore::Damage::cellIndexForRect const):
(WebCore::Damage::unite):
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, AddRects)):

Canonical link: https://commits.webkit.org/293001@main
